### PR TITLE
*: improve ErrKeyExists generation and MemoizeStr 

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve?

Issue Number: ref #53004

Problem Summary:

### What changed and how does it work?

- `MemoizeStr` does not cache the computed result after https://github.com/pingcap/tidb/pull/11450, this PR reverts it.
- Add `GenKeyExistsErr` to ensure using `[]string` as argument

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by hawkingrei — APPROVED** [review]

File: pkg/table/tables/tables.go [hunk 1]
```diff
@@ -1165,11 +1165,11 @@ func (t *TableCommon) addIndices(sctx table.MutateContext, recordID kv.Handle, r
 		if !skipCheck && v.Meta().Unique {
 			// Make error message consistent with MySQL.
 			tablecodec.TruncateIndexValues(t.meta, v.Meta(), indexVals)
-			entryKey, err := genIndexKeyStr(indexVals)
+			entryKey, err := genIndexKeyStrs(indexVals)
```
Comment (D3Hunter) [inline pkg/table/tables/tables.go:1168]:
```suggestion
			colStrVals, err := genIndexKeyStrs(indexVals)
```

Comment (D3Hunter) [reply pkg/table/tables/tables.go:1168]: there are a few in below code, can change them together

**Review by D3Hunter — APPROVED** [review]

**Review by Benjamin2037 — APPROVED** [review]
LGTM

**Pull request merged**